### PR TITLE
refactor preprocessing resampling, validation, and alignment

### LINKED ISSUE #556: accounting for marginal nighttime and clearsky conditions when resampling observations
Let's say we want to compare 1 minute observations of GHI to hourly forecasts of GHI, but exclude nighttime data. We define nighttime as solar zenith > 87 degrees at the time label (regardless of interval length - a separate issue).

We discard any "nighttime" values from the observations series, resample the results to hourly average, and compute the metrics. If 90% or more of 1 minute values are nighttime then we set the interval to NaN, and it will be dropped from the comparison.

The problem is that our procedure requires that forecasters use the same definition of nighttime. Even for a perfect forecast, a forecaster than defines nighttime as solar zenith > 90 degrees will compute and submit a slightly different hourly average than we'll calculate from observations.

I don't see a way out of this other than to exclude any interval that contains a single nighttime point.

### PR DESCRIPTION
**Pull request by wholmgren**
- [x] Closes #556  .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

There is a problem with the current implementation. Flags that represent bad data should be excluded from the time series before resampling (e.g. limits exceeded) and that's not done here. We're trading one problem for another. I don't know how to solve this without a richer flag datamodel or awful hacks. So maybe it's worth addressing that now. Or we can say this is a net win, move on, and address that later as originally planned. I guess I would say address it now but I don't have a good idea of what the datamodel changes would actually look like.

@alorenzo175 can you review and let me know what you think?

### CODE REVIEW

**Review by alorenzo175 — COMMENTED** [review]
Looks good so far. I think the datamodel changes basically involve adding keys like threshold_percentage to QualityFlagFilter. To handle this new issue, I think a key like discard_before_resample would help. 

We should also keep in mind that one can specify multiple quality flag filters in the report. Right now they are merged into one (https://github.com/SolarArbiter/solarforecastarbiter-core/pull/580/files#diff-715b43f6fac14c48d8729da1c197d286R430) but if we go the route of adding these keys, maybe the filters should be applied successively.

**Comment by wholmgren** [comment]
> I think the datamodel changes basically involve adding keys like threshold_percentage to QualityFlagFilter. To handle this new issue, I think a key like discard_before_resample would help.

Would these be exposed in the report json or would they be preset for predefined QualityFlagFilters for each kind of quality flag?

**Comment by alorenzo175** [comment]
> > I think the datamodel changes basically involve adding keys like threshold_percentage to QualityFlagFilter. To handle this new issue, I think a key like discard_before_resample would help.
> 
> Would these be exposed in the report json or would they be preset for predefined QualityFlagFilters for each kind of quality flag?

I was thinking they would be in the report json. If we wanted some kind of presets, I would think the dashboard could handle that like it might already do for other fields.

File: solarforecastarbiter/datamodel.py [hunk 1]
```diff
@@ -1379,17 +1379,19 @@ class QualityFlagFilter(BaseFilter):
         ``DERIVED_MASKS`` keys.
         These periods will be excluded from the analysis.
     """
-    quality_flags: Tuple[str, ...] = (
-        'UNEVEN FREQUENCY', 'LIMITS EXCEEDED', 'CLEARSKY EXCEEDED',
-        'DAYTIME STALE VALUES', 'INCONSISTENT IRRADIANCE COMPONENTS'
-    )
+    type: str
```
Comment (alorenzo175) [inline solarforecastarbiter/datamodel.py:1382]: It isn't ideal, but we should probably keep the `quality_flags` key (and the other parameters would apply to all flags in that tuple). Otherwise all current reports would be broken since I think the report json is loaded through the report datamodel before displaying/downloading.

Comment (wholmgren) [reply solarforecastarbiter/datamodel.py:1382]: So keep the old `quality_flags` key in addition to the new keys? I agree that could work. A couple of things to consider as alternatives:

1. script to migrate current reports from old json to new json
2. create new objects and json keys, leave the old ones untouched until some later time when we implement 1

I'm not sure what would be less work because the testing implications for keeping the old keys sound miserable.

Comment (alorenzo175) [reply solarforecastarbiter/datamodel.py:1382]: Well keep old key, and add new keys except `type`. So all flags in `quality_flags` are processed with the other options. You would probably have the same options for many of the flags anyway, and you can still have multiple QualityFlagFilters with different sets of flags.

We've done the migration before an it was a real pain (and motivation to eventually get the report schema versioned). One problem is you also have to modify the json core posts to the API. 

I don't think testing would be bad. All the keys are still used (if `type` is removed), we just have a couple of new keys w/ reasonable defaults if empty

**Comment by wholmgren** [comment]
Should these have the same behavior?

```python
# separate flags:
qflag_1 = QualityFlagFilter(('NIGHTTIME', ), discard_before_resample=False, resample_threshold_percentage=30.)
qflag_2 = QualityFlagFilter(('CLEARSKY', ), discard_before_resample=False, resample_threshold_percentage=30.)

# combined flags
qflag_combined = QualityFlagFilter(('NIGHTTIME', 'CLEARSKY'), discard_before_resample=False, resample_threshold_percentage=30.)
```

Imagine a 15 minute validation time series like

```
nighttime = [True, False, False, False]
clearsky = [False, False, True, False]
```

and we want to resample into 1 hour. If we apply the filters individually, we keep the hourly interval. But if we combine the filters we have `[True, False, True, False]` and we discard the interval. 

I think the "separate flags" case should keep the interval and the "combined flags" case should discard the interval, but I'm open to the idea that they should have the same behavior.

**Comment by alorenzo175** [comment]
Interesting. I guess I always thought they would have the same behaviour and be treated separately. I can see how the combined flags case could be useful. Just need to make sure it's documented well.

**Comment by wholmgren** [comment]
More scope creep.... `TimeOfDayFilter` and `ValueFilter` are not currently implemented, but they'll need to fit into this pattern eventually. It's not obvious to me how that would work. I feel like they might need to be eliminated and instead set as keys on `QualityFlagFilter` (perhaps renamed). It's worth considering that before committing to this refactor.

**Comment by wholmgren** [comment]
`test_preprocessing.py` passes. Remaining to do:

- [ ] Add robust tests for `filter_resample`/`_resample_obs`
- [ ] Update fixtures for new json representation
- [ ] Fix whatever tests are still broken after updating fixtures
- [ ] Determine if there are complicating factors from anticipated implementations of `TimeOfDayFilter` and `ValueFilter`.
- [ ] Create issue in `solarforecastarbiter-api`?

File: solarforecastarbiter/metrics/preprocessing.py [hunk 2]
```diff
@@ -307,15 +357,17 @@ def filter_resample(fx_obs, fx_data, obs_data, ref_data, quality_flag):
     ref_fx = fx_obs.reference_forecast
 
     # raise ValueError if intervals don't match
+    # move this to align?
```
Comment (alorenzo175) [inline solarforecastarbiter/metrics/preprocessing.py:360]: If anything, I would make `_check_ref_fx` public and run that in `process_forecast_observations` early on (catching and logging as appropriate)

File: solarforecastarbiter/metrics/preprocessing.py [hunk 3]
```diff
@@ -213,47 +223,66 @@ def _resample_obs(obs, fx, obs_data, quality_flags):
...
+        threshold = f.resample_threshold_percentage / 100. * interval_ratio
+        flagged = resampled_flags_count > threshold
+        to_discard |= flagged
+        counts[filter_name] = flagged.sum()
 
     # determine intervals with too many pre-resampling points removed
+    # this is equivalent to the old apply_validation + exlude pattern, but
+    # not sure we actually want to do this in our new paradigm.
+    # and should it always be 10%? should it be controlled by
+    # resampled_threshold_percentage on a filter by filter basis?
```
Comment (alorenzo175) [inline solarforecastarbiter/metrics/preprocessing.py:264]: I think so

File: solarforecastarbiter/metrics/preprocessing.py [hunk 4]
```diff
@@ -213,47 +223,66 @@ def _resample_obs(obs, fx, obs_data, quality_flags):
...
-    discard_after_resample_flags = [
-        f for f in quality_flags if not f.discard_before_resample
-    ]
-    for flag in discard_after_resample_flags:
-        threshold = flag.resample_threshold_percentage * interval_ratio
-        flagged = resampled_flags_count[flag.type] > threshold
-        to_discard += flagged
-        counts[flag.type] = flagged.sum()
+
+    for f in filter(lambda x: not x.discard_before_resample, quality_flags):
```
Comment (alorenzo175) [inline solarforecastarbiter/metrics/preprocessing.py:244]: This function is really hard to follow. Can it be split up into `_discard_before_resample` and `_discard_after_resample`? would probably need a final join on the counts and .loc on the series to drop the times from `_discard_before_resample`

**Review by alorenzo175 — COMMENTED** [review]
From what I understand, it is getting closer.

> More scope creep.... TimeOfDayFilter and ValueFilter are not currently implemented, but they'll need to fit into this pattern eventually. It's not obvious to me how that would work. I feel like they might need to be eliminated and instead set as keys on QualityFlagFilter (perhaps renamed). It's worth considering that before committing to this refactor.

Why would they need to be combined w/ QualityFlagFilter? Can this all be more modular? Like `_filter_qualityflag_before`, `_filter_qualityflag_after`, `_filter_time_before`, etc that take the series and a single filter object and return a boolean series w/ the resampled index and the counts. Then each filter is applied in the order of the report filters tuple. We might have to decide to just dump the nans first or something. Maybe there is something else I'm missing that you would want to take it all into account at the same time.

>  Create issue in solarforecastarbiter-api?

sure, but might currently work w/o adjustments since the schema just expects a list of dicts

**Comment by wholmgren** [comment]
> Why would they need to be combined w/ QualityFlagFilter?

It's an extension of the problem discussed in https://github.com/SolarArbiter/solarforecastarbiter-core/pull/580#issuecomment-704525057. Now imagine you have 15 minute pseudoflags like

```
nighttime = [True, False, False, False]
clearsky = [False, True, False, False]
value = [False, False, True, False]
time = [False, False, False, True]
```

And you would discard the interval if 75% of the points were flagged. If you compute the threshold on each flag and then compute the OR, you keep the interval. If you compute the joined flag, and then compute the OR, you discard the interval. If you want to allow for the possibility of the latter, then I don't see how you separate the quality flag, time, and value filter operations.

**Comment by alorenzo175** [comment]
I would say we should keep it a little simpler then and not allow the combination. Although, I guess it wouldn't be too bad if there were just a Filter object that had keys `quality_flags`, `time_of_day_range`, `value_range`. Either way, I think we can lay the groundwork in this PR and make a follow on soon after if necessary

File: solarforecastarbiter/metrics/preprocessing.py (hunk 2 above)
Comment (wholmgren) [reply solarforecastarbiter/metrics/preprocessing.py]: `preprocessing.check_reference_forecast_consistency`

File: solarforecastarbiter/metrics/preprocessing.py (hunk 4 above)
Comment (wholmgren) [reply solarforecastarbiter/metrics/preprocessing.py:344]: 8a44022

File: solarforecastarbiter/metrics/preprocessing.py (hunk 3 above)
Comment (wholmgren) [reply solarforecastarbiter/metrics/preprocessing.py]: c862e55

**Comment by wholmgren** [comment]
I'd like to think it makes more sense now. 

> Although, I guess it wouldn't be too bad if there were just a Filter object that had keys quality_flags, time_of_day_range, value_range. Either way, I think we can lay the groundwork in this PR and make a follow on soon after if necessary

I added a few comments on where we could easily add support for this. Best left to another PR, though.

File: solarforecastarbiter/datamodel.py [hunk 5]
```diff
@@ -1369,20 +1369,38 @@ def from_dict(model, input_dict, raise_on_extra=False):
...
         ``DERIVED_MASKS`` keys.
         These periods will be excluded from the analysis.
+    discard_before_resample : bool, default True
+        Determines if points should be discarded before resampling or
+        during resampling (when ``resample_threshold_percentage`` is
+        exceeded).
+    resample_threshold_percentage : float, default 10.
+        The percentage of points in a resampled interval that must be
+        flagged for the resampled interval to be flagged.
+        Ignored if ``discard_before_resample`` is True.
```
Comment (alorenzo175) [inline solarforecastarbiter/datamodel.py:1387]: I think this needs updating?

**Review by alorenzo175 — APPROVED** [review]
